Feature/extend cluster-agent REST API

iskitsas opened the pull request:
Expose more APIs from cluster agent to support manage and view operations for MicroK8s, for easy integration with other apps, like dashboards.

The exposed APIs will help to integrate the [MicroK8s Web Console](https://github.com/sastix/microk8s-addons) ([microdash](https://hub.docker.com/repository/docker/sastix/microdash)) as an AddOn.

Along with the new APIs, I am enabling Swagger-UI as a complete documentation page and easy testing web tool for the endpoints. The page is exposed under https://127.0.0.1:25000/swagger

Attaching some screenshots:
![microk8s-swaggerui-home](https://user-images.githubusercontent.com/1267871/77917589-6a7c7b00-72a3-11ea-8970-2e42563623b9.png)
![microk8s-swaggerui-version](https://user-images.githubusercontent.com/1267871/77917607-6f412f00-72a3-11ea-8598-b723640fb2b5.png)
![microk8s-swaggerui-version-tryout](https://user-images.githubusercontent.com/1267871/77917612-6fd9c580-72a3-11ea-9f8f-5758cca00c27.png)
![microk8s-swaggerui-configure](https://user-images.githubusercontent.com/1267871/77917582-68b2b780-72a3-11ea-96fc-74a9079fcafe.png)
![microk8s-swaggerui-get-services](https://user-images.githubusercontent.com/1267871/77917584-694b4e00-72a3-11ea-8176-d85a8c71587a.png)
![microk8s-swaggerui-overview-tryout](https://user-images.githubusercontent.com/1267871/77917592-6bada800-72a3-11ea-84a9-d2c790a1b5ea.png)
![microk8s-swaggerui-addon-enable](https://user-images.githubusercontent.com/1267871/77917576-66e8f400-72a3-11ea-8705-ee9b0720ca68.png)
![microk8s-swaggerui-service-logs](https://user-images.githubusercontent.com/1267871/77917597-6cded500-72a3-11ea-92c2-cb86a3d854dc.png)
![microk8s-swaggerui-service-stop](https://user-images.githubusercontent.com/1267871/77917604-6ea89880-72a3-11ea-9bec-e813d61c08f5.png)

balchua commented:
Quick question, can `/status` and `/version` be `GET` requests?  I don't see any security implications on these.  Unless i missed it.

iskitsas commented:
> I like this PR, but i do have a quick question, can `/status` and `/version` be `GET` requests? I don't see any security implications on these. Unless i missed it.

Yes it is very easy. Just keep in mind that we send a body with the POST request where we have the callback token for validation, eg:

```
{
  "callback": "xyztoken"
}
```

It is not common to send body with the GET request (it is possible though). The callback token could be replaced with a query parameter, but this will **expose the token on the url**, eg:

```
https://127.0.0.1:25000/cluster/api/v1.0/version?cbtoken=xyztoken
```

OR 

we could remove the token validation for those 2 calls.

So it is just decision we have to make :)

balchua commented:
Im thinking for `GET` method there is no need to pass tokens.  @ktsakalozos wdyt?

balchua commented:
I have another request if possible.  What about a secure method to get the kubeconfig too. 😁

iskitsas commented:
> I have another request if possible. What about a secure method to get the kubeconfig too.

Yes, it is possible :)

iskitsas commented:
> > I have another request if possible. What about a secure method to get the kubeconfig too.
> 
> Yes, it is possible :)
Hi @balchua, @ktsakalozos, 

I have updated the PR by adding a new POST **/config** endpoint for getting microk8s.config info and expose **/version** and **/status** under a GET request, with no need to pass the callback token. 

In order to enable those, **you must create a file** with the name 'enable-extended-api' under '/var/snap/microk8s/current/args/enable-extended-api', HAVING as content THE SAME TOKEN VALUE with '/var/snap/microk8s/current/credentials/callback-token.txt'. 

For the POST /config the user must also send the callback token in the body of the request as json (like we are doing for the other POST endpoints), and for the GET /version and /status the user does not need to send it.

I have also updated the README with full examples to test it from a shell with curl, and swaggerUI so you can try them from browse.

Some screenshots from Swagger UI with new/updated endpoints:

Home:
![swagger-ui-home](https://user-images.githubusercontent.com/1267871/78774471-63d5be00-799d-11ea-8297-d45bd82cd45e.png)

GET-version-with-disabled-extended-api:
![swagger-ui-GET-version-with-disabled-extended-api](https://user-images.githubusercontent.com/1267871/78774405-4d2f6700-799d-11ea-8b5d-7d249aff59ac.png)

GET-version-with-enabled-extended-api:
![swagger-ui-GET-version-with-enabled-extended-api](https://user-images.githubusercontent.com/1267871/78774427-54567500-799d-11ea-8494-ef3fae103339.png)

GET-status-with-enabled-extended-api:
![swagger-ui-GET-status-with-enabled-extended-api](https://user-images.githubusercontent.com/1267871/78774512-718b4380-799d-11ea-919f-d595485d60d0.png)

GET-config-with-enabled-extended-api:
![swagger-ui-GET-config-with-enabled-extended-api](https://user-images.githubusercontent.com/1267871/78774543-7b14ab80-799d-11ea-958c-e74a4ecd03e4.png)

balchua commented:
Thanks @iskitsas.  The more i think about this, i was thinking we should put the token as part of the header, just like what most token based authn/authz systems do.
For example `Authorization Bearer:`

This way it is easy to determine which methods do not pose side effects like `GET`.

What do you guys think?

iskitsas commented:
> Thanks @iskitsas. The more i think about this, i was thinking we should put the token as part of the header, just like what most token based authn/authz systems do.
> For example `Authorization Bearer:`
> 
> This way it is easy to determine which methods do not pose side effects like `GET`.
> 
> What do you guys think?

Can easily be changed to header based checking :)

iskitsas commented:
> > Thanks @iskitsas. The more i think about this, i was thinking we should put the token as part of the header, just like what most token based authn/authz systems do.
> > For example `Authorization Bearer:`
> > This way it is easy to determine which methods do not pose side effects like `GET`.
> > What do you guys think?
> 
> Can easily be changed to header based checking :)

Just added another way to pass the token. You can now use the header "Callback-Token: <value>" to validate

ktsakalozos commented on scripts/cluster/agent.py:
```diff
@@ -423,6 +443,285 @@ def configure():
+    return output
+
+
+@app.route('/{}/start'.format(CLUSTER_API), methods=['POST'])
```
Given you will need to start MicroK8s from the command line, I do not think we should allow the stopping of services via the REST api. It would be possible users may be surprised by this.

ktsakalozos replied:
Also, I understand you know why this call is not going to be called.

iskitsas replied:
Will remove it

ktsakalozos commented on scripts/cluster/agent.py:
```diff
@@ -423,6 +443,285 @@ def configure():
+    return output
+
+
+@app.route('/{}/overview'.format(CLUSTER_API), methods=['POST'])
```
Getting access tot he cluster like this is not something we recommend. If you want to have an overview of the cluster you can query the cluster's API server directly.

ktsakalozos replied:
The cluster agent API is not meant to be a proxy for the k8s API server.

MicroDash runs as a pod in MicroK8s and thus should access to the cluster API server as described in the official docs [1]. This way we comply with the authn & authz policies set by the cluster administrator.

[1] https://kubernetes.io/docs/tasks/administer-cluster/access-cluster-api/#accessing-the-api-from-within-a-pod

ktsakalozos commented on scripts/cluster/agent.py:
```diff
@@ -423,6 +443,285 @@ def configure():
+    return output
+
+
+@app.route('/{}/status'.format(CLUSTER_API), methods=['POST', 'GET'])
```
I am not sure about the GET part. Yes it makes sense to be a get but you need to pass the callback token in a secure way.

balchua replied:
Just my 2 cents, would passing the token through header be acceptable?
The same thing as what most APIs are doing like oauth2.
Wdyt?

iskitsas replied:
The way it is implemented now, is allowing GET request with no token passed, if and only if 'enable-extended-api' is true.

Additionally, passing the token through Header or a json body, both ways are supported. Token is never exposed though a query param

ktsakalozos commented on scripts/cluster/agent.py:
```diff
@@ -423,6 +443,285 @@ def configure():
+    return app.response_class(response=json.dumps(json_output, sort_keys=False, indent=4), status=200, mimetype='application/json')
+
+
+@app.route('/{}/config'.format(CLUSTER_API), methods=['POST'])
```
I do not think you need this for MicroDash. Please make a case why this is needed in a separate PR.

balchua replied:
I may have a use case if this method returns the kubeconfig? 
If yes, it can be helpful when working with clusterapi when u have to return the kubeconfig through kubernetes secret.

iskitsas replied:
This is only accessible if you 'enable-extended-api', passing also the callback-token using a Header or the json body

ktsakalozos commented on scripts/cluster/agent.py:
```diff
@@ -423,6 +443,285 @@ def configure():
+    if 'Content-Type' in request.headers and request.headers['Content-Type'] == 'application/json':
+        callback_token = request.json['callback']
+    elif 'Callback-Token' in request.headers:
+        callback_token = request.headers['Callback-Token']
```
Does this need to be part of this PR? Could you put this in a saparate PR so we can discuss with @balchua the name of the header we would like to pass the token in?

iskitsas replied:
I could make a separate PR, however It is just 2 lines of code and a decision for the name. I would propose @balchua to have a look, and if still is the need for a separate PR, I can do it

ktsakalozos commented on scripts/cluster/agent.py:
```diff
@@ -423,6 +443,285 @@ def configure():
+    return {"valid": valid, "response": resp}
+
+
+def is_extended_api_enabled():
```
Same, here does this need to be part of the MicroDash PR? I understand this is a flag to enable/disable certain functionalities. I think this belongs into its own PR so we discuss on what we want to feature-flag and how. Thank you.

iskitsas replied:
will do in a separate PR

ktsakalozos commented on scripts/cluster/agent.py:
```diff
@@ -423,6 +443,285 @@ def configure():
+    return output
+
+
+@app.route('/{}/service/logs'.format(CLUSTER_API), methods=['POST'])
```
Could you put the log collection under the status endpoint? Thank you.

iskitsas replied:
Do you mean /status/service/logs? Or extend the json body which is sent to /status, to support the service logs with new properties?

eg on /status:
`{"callback":"xyztoken", "servicelogs":{"name":"microk8s.daemon-flanneld", "lines":30}}

iskitsas commented on scripts/cluster/agent.py:
```diff
@@ -423,6 +443,285 @@ def configure():
+    return json_output["installed"].split(' ')[0]
+
+
+@app.route('/{}/services'.format(CLUSTER_API), methods=['POST'])
```
will remove it from this PR and make a separate, with the suggestions for making a cleaner, more readable and domain specific API

iskitsas commented on scripts/cluster/agent.py:
```diff
@@ -12,18 +13,37 @@
+                        "a file with the name 'enable-extended-api' under '" + enable_extended_api_file + "', " \
+                        "HAVING THE SAME TOKEN VALUE with '" + callback_token_file + "'"
+
+# -- swagger specific --
```
Will introduce a new approach on separate PR
